spot-ml/src/main/scala/org/apache/spot/SpotSparkLDAWrapper.scala

```diff
@@ -34,43 +35,59 @@ object SpotSparkLDAWrapper {
              ldaBeta: Double = 1.001,
              maxIterations: Int = 20): SpotLDAOutput = {
 
+    import sqlContext.implicits._
+
+    val docWordCountCache = docWordCount.cache()
+
+    // Forcing an action to cache results.
+    docWordCountCache.count()
+
     // Create word Map Word,Index for further usage
     val wordDictionary: Map[String, Int] = {
-      val words = docWordCount
-        .cache
+      val words = docWordCountCache
         .map({ case SpotLDAInput(doc, word, count) => word })
         .distinct
         .collect
       words.zipWithIndex.toMap
     }
 
-    val distinctDocument: Array[String] = docWordCount.map({ case SpotLDAInput(doc, word, count) => doc }).distinct.collect
+    val documentDictionary: DataFrame = docWordCountCache
+      .map({ case SpotLDAInput(doc, word, count) => doc })
+      .distinct
+      .zipWithIndex
+      .toDF(DocumentName, DocumentId)
+      .cache
 
-    // Create document Map Index, Document for further usage
-    val documentDictionary: Map[Int, String] = {
-      distinctDocument
-        .zipWithIndex
-        .sortBy(_._2)
-        .map(kv => (kv._2, kv._1))
-        .toMap
-    }
-    val docStrToID: Map[String, Int] = documentDictionary.map(_.swap)
+    // Forcing an action to cache results
+    documentDictionary.count()
 
     //Structure corpus so that the index is the docID, values are the vectors of word occurrences in that doc
-    val ldaCorpus: RDD[(Long, Vector)] = formatSparkLDAInput(docWordCount, docStrToID, wordDictionary)
+    val ldaCorpus: RDD[(Long, Vector)] =
+      formatSparkLDAInput(docWordCountCache,
+      documentDictionary,
+      wordDictionary,
+      sqlContext)
+
+    docWordCountCache.unpersist()
 
     //Instantiate optimizer based on input
     val optimizer = ldaOptimizer match {
       case "em" => new EMLDAOptimizer
-      case "online" => new OnlineLDAOptimizer().setMiniBatchFraction(0.05 + 1.0 / distinctDocument.size)
+      case "online" => new OnlineLDAOptimizer().setMiniBatchFraction(0.05 + 1.0 / documentDictionary.count)

```

Comment:
I know Julia wrote this code... but I'm not sure that we should handle parameterization of the optimizer....  I couldn't get the online optimizer to work with code below